### PR TITLE
feat: implement SQL lite storage

### PR DESCRIPTION
**Pull request by Stefan-Ethernal**
### What does this PR do?
This PR introduces SQL lite storage to the zkevm-eth-txmanager, in case the `PersistenceFilename` config param is provided.

### Reviewers

Main reviewers:
- @ToniRamirezM 
- @goran-ethernal

### CODE REVIEW

File: common/common.go [hunk 1]
```diff
@@ -1,8 +1,37 @@
...
+func BoolToInteger(v bool) int {
+	if v {
+		return 1
+	}
+
+	return 0
+}
+
+// ToAddressOrNil converts a string to a common.Address pointer or returns nil if empty.
+func ToAddressOrNil(addr string) *common.Address {
```
Comment (goran-ethernal) [inline common/common.go:25]:
```suggestion
func ToAddressPointer(addr string) *common.Address {
```

File: ethtxmanager/sqlstorage/sqlstorage.go [hunk 2]
```diff
@@ -0,0 +1,479 @@
...
+
+	return &SqlStorage{db: db}, nil
+}
+
+// Add persist a monitored transaction into the SQL database.
+func (s *SqlStorage) Add(ctx context.Context, mTx types.MonitoredTx) error {
+	mTx.CreatedAt = time.Now()
+	mTx.UpdatedAt = mTx.CreatedAt
+
+	query := `
```
Comment (goran-ethernal) [inline ethtxmanager/sqlstorage/sqlstorage.go:84]: Maybe take a look at the `meddler` we used on `cdk` repo, to handle inserting and querying structs from sqlite?

File: common/common.go (hunk 1 above)
Comment (Stefan-Ethernal) [reply common/common.go]: Fixed https://github.com/0xPolygon/zkevm-ethtx-manager/pull/72/commits/f3fd2802e458635c61ca57d3159199e750a8cc5a

**Review by joanestebanr — COMMENTED** [review]
I'm wondering if make sense to support dbTx?

File: ethtxmanager/sqlstorage/sqlstorage.go [hunk 3]
```diff
@@ -0,0 +1,271 @@
...
+// It first opens a connection to the SQLite database and then runs the necessary migrations.
+// If any error occurs during the database connection or migration process, it returns an error.
+func NewStorage(driverName, dbPath string) (*SqlStorage, error) {
+	db, err := sql.Open(driverName, dbPath)
+	if err != nil {
+		return nil, err
+	}
+
+	_, err = db.Exec(`
+		pragma journal_mode = WAL;
```
[the remark this answers is not in the document]
Comment (Stefan-Ethernal) [reply ethtxmanager/sqlstorage/sqlstorage.go:79]: The schema contains only a single table, so we don't need it (at least for now), but I'm going to add it, suppose it doesn't harm.

File: ethtxmanager/sqlstorage/sqlstorage.go [hunk 4]
```diff
@@ -0,0 +1,271 @@
...
+
+	// baseDeleteStatement represents the base delete statement that deletes all the records from the monitored_txs table
+	baseDeleteStatement = "DELETE FROM monitored_txs"
+
+	// monitoredTxsTable is table name for persisting MonitoredTx objects
+	monitoredTxsTable = "monitored_txs"
+)
+
+var (
+	errNoRowsInResultSet = errors.New("sql: no rows in result set")
```
[the remark this answers is not in the document]
Comment (Stefan-Ethernal) [reply ethtxmanager/sqlstorage/sqlstorage.go:59]: But this error message is coming from the meddler. I have introduced it in order to treat this use case differently (unfortunatelly, I haven't seen there is some constant that I'm able to reuse and test against it)...

File: ethtxmanager/sqlstorage/sqlstorage.go [hunk 5]
```diff
@@ -0,0 +1,271 @@
...
+		query += ")"
+	}
+
+	// Add ordering by creation date (oldest first)
+	query += " ORDER BY created_at ASC"
+
+	// Use meddler.QueryAll to retrieve the monitored transactions
+	var transactions []*types.MonitoredTx
+	if err := meddler.QueryAll(s.db, &transactions, query, args...); err != nil {
+		return nil, fmt.Errorf("failed to get monitored transactions: %w", err)
```
[the remark this answers is not in the document]
Comment (Stefan-Ethernal) [reply ethtxmanager/sqlstorage/sqlstorage.go]: Agree, fixed https://github.com/0xPolygon/zkevm-ethtx-manager/pull/72/commits/8cbeceace4c138d2874caceeaacf91984abdc813

File: ethtxmanager/sqlstorage/sqlstorage.go [hunk 6]
```diff
@@ -0,0 +1,271 @@
...
+	if rowsAffected == 0 {
+		return types.ErrNotFound
+	}
+
+	return nil
+}
+
+// Get retrieves a monitored transaction from the database by its ID.
+// If the transaction is not found, it returns an ErrNotFound error.
+func (s *SqlStorage) Get(ctx context.Context, id common.Hash) (types.MonitoredTx, error) {
```
[the remark this answers is not in the document]
Comment (Stefan-Ethernal) [reply ethtxmanager/sqlstorage/sqlstorage.go]: Agree, fixed https://github.com/0xPolygon/zkevm-ethtx-manager/pull/72/commits/8cbeceace4c138d2874caceeaacf91984abdc813

File: ethtxmanager/sqlstorage/sqlstorage.go [hunk 7]
```diff
@@ -0,0 +1,271 @@
...
+		}
+	}
+
+	return err
+}
+
+// Remove deletes a monitored transaction from the database by its ID.
+// If the transaction does not exist, it returns an ErrNotFound error.
+func (s *SqlStorage) Remove(ctx context.Context, id common.Hash) error {
+	query := fmt.Sprintf("%s WHERE id = $1", baseDeleteStatement)
```
[the remark this answers is not in the document]
Comment (Stefan-Ethernal) [reply ethtxmanager/sqlstorage/sqlstorage.go]: Agree, fixed https://github.com/0xPolygon/zkevm-ethtx-manager/pull/72/commits/8cbeceace4c138d2874caceeaacf91984abdc813

File: ethtxmanager/sqlstorage/sqlstorage.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (Stefan-Ethernal) [reply ethtxmanager/sqlstorage/sqlstorage.go:81]: Fixed https://github.com/0xPolygon/zkevm-ethtx-manager/pull/72/commits/8cbeceace4c138d2874caceeaacf91984abdc813

**Comment by Stefan-Ethernal** [comment]
> I'm wondering if make sense to support dbTx?

Can you elaborate some more? Do you mean to be able to do multiple changes in the same transaction (basically to ensure transactional mode)?

File: ethtxmanager/sqlstorage/sqlstorage.go [hunk 8]
```diff
@@ -0,0 +1,271 @@
...
+		return nil, err
+	}
+
+	initMeddler()
+
+	return &SqlStorage{db: db}, nil
+}
+
+// Add persist a monitored transaction into the SQL database.
+func (s *SqlStorage) Add(ctx context.Context, mTx types.MonitoredTx) error {
```
[the remark this answers is not in the document]
Comment (Stefan-Ethernal) [reply ethtxmanager/sqlstorage/sqlstorage.go]: Fixed https://github.com/0xPolygon/zkevm-ethtx-manager/pull/72/commits/8cbeceace4c138d2874caceeaacf91984abdc813

File: ethtxmanager/sqlstorage/sqlstorage.go [hunk 9]
```diff
@@ -0,0 +1,271 @@
...
+		}
+
+		args = append(args, *toBlock)
+	}
+
+	// Use meddler.QueryAll to execute the query and scan into the result slice.
+	var monitoredTxs []*types.MonitoredTx
+	err := meddler.QueryAll(s.db, &monitoredTxs, query, args...)
+	if err != nil {
+		return nil, fmt.Errorf("failed to query monitored transactions: %w", err)
```
[the remark this answers is not in the document]
Comment (Stefan-Ethernal) [reply ethtxmanager/sqlstorage/sqlstorage.go]: Fixed https://github.com/0xPolygon/zkevm-ethtx-manager/commit/8cbeceace4c138d2874caceeaacf91984abdc813

File: ethtxmanager/sqlstorage/sqlstorage.go [hunk 10]
```diff
@@ -0,0 +1,271 @@
...
+
+	err := meddler.Insert(s.db, monitoredTxsTable, &mTx)
+	if err != nil {
+		sqlErr, success := UnwrapSQLiteErr(err)
+		if !success {
+			return err
+		}
+
+		if sqlErr.Code == sqlite.ErrConstraint {
+			return fmt.Errorf("transaction with ID %s already exists", mTx.ID)
```
[the remark this answers is not in the document]
Comment (Stefan-Ethernal) [reply ethtxmanager/sqlstorage/sqlstorage.go:112]: Can you elaborate a bit more?

File: ethtxmanager/sqlstorage/migrations/0001.sql [hunk 11]
```diff
@@ -0,0 +1,31 @@
+-- +migrate Up
+CREATE TABLE IF NOT EXISTS monitored_txs (
+    id CHAR(66) PRIMARY KEY,
+    from_address CHAR(42) NOT NULL,
+    to_address CHAR(42),
+    nonce BIGINT NOT NULL,
+    "value" TEXT,                  -- *big.Int
```
[the remark this answers is not in the document]
Comment (Stefan-Ethernal) [reply ethtxmanager/sqlstorage/migrations/0001.sql:7]: I thought that `BIGINT` has a predefined range of values, whereas `*big.Int` can be arbitrary value, and therefore I thought it is safer to use the `TEXT` to store the values. WDYT?

File: ethtxmanager/sqlstorage/migrations/0001.sql [hunk 12]
```diff
@@ -0,0 +1,31 @@
...
+    gas_offset BIGINT,
+    gas_price TEXT,                -- *big.Int
+    blob_sidecar BLOB,
+    blob_gas BIGINT,
+    blob_gas_price TEXT,           -- *big.Int
+    gas_tip_cap TEXT,              -- *big.Int
+    "status" TEXT NOT NULL,
+    block_number TEXT NOT NULL,    -- *big.Int
+    history TEXT,                  -- JSON encoded string
+    created_at TEXT NOT NULL,
```
[the remark this answers is not in the document]
Comment (Stefan-Ethernal) [reply ethtxmanager/sqlstorage/migrations/0001.sql:19]: I thought that SQL lite does not support it (https://www.sqlite.org/datatype3.html). Let me try that out...

File: ethtxmanager/sqlstorage/sqlstorage.go [hunk 13]
```diff
@@ -0,0 +1,271 @@
...
+		return types.MonitoredTx{}, err
+	}
+
+	return mTx, nil
+}
+
+// GetByStatus retrieves monitored transactions from the database that match the provided statuses.
+// If no statuses are provided, it returns all transactions.
+// The transactions are ordered by their creation date (oldest first).
+func (s *SqlStorage) GetByStatus(ctx context.Context, statuses []types.MonitoredTxStatus) ([]types.MonitoredTx, error) {
```
[the remark this answers is not in the document]
Comment (Stefan-Ethernal) [reply ethtxmanager/sqlstorage/sqlstorage.go]: Fixed https://github.com/0xPolygon/zkevm-ethtx-manager/commit/8cbeceace4c138d2874caceeaacf91984abdc813

File: ethtxmanager/sqlstorage/sqlstorage.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (joanestebanr) [reply ethtxmanager/sqlstorage/sqlstorage.go:112]: I mean create a var to use as check the result in the usage side... something like: 
```
var (
    ErrDuplicatedKeys = ...
)
```

File: ethtxmanager/sqlstorage/migrations/0001.sql (hunk 11 above)
[the remark this answers is not in the document]
Comment (joanestebanr) [reply ethtxmanager/sqlstorage/migrations/0001.sql:7]: Do you known that is the data stored on 'value' field? my point is if you set numerical values as TEXT are not able to do selects like `value >= 1234`... I try to find out the limits of `BIGINT` type but in sqlite documentation is not clear

**Comment by joanestebanr** [comment]
> > I'm wondering if make sense to support dbTx?
> 
> Can you elaborate some more? Do you mean to be able to do multiple changes in the same transaction (basically to ensure transactional mode)?
> 
> IMO, probably doesn't make too much sense, because it is only a single table, but I'm not sure (it depends of the usage). There might be some valid use case for it 🤔

I don't known that is done internally but even with just 1 table make sense: 
- You do multiples INSERTS but want to rollback if any of them fails
- You do some INSERTS and check something over them in another function (that need to be in the same set of data)
Basically, it depends how works internally. But in the future make sense, that if we modify it, we think in that model because is the common in the rest of Database usage

File: ethtxmanager/sqlstorage/migrations/0001.sql (hunk 11 above)
[the remark this answers is not in the document]
Comment (Stefan-Ethernal) [reply ethtxmanager/sqlstorage/migrations/0001.sql:7]: If we are confident that neither of those big ints are going to be larger than `int64` range and which I'm pretty confident about, although there is no trace about it in the SQL docs (https://dba.stackexchange.com/a/125147), then I can change it to `BIGINT` type. Also, I encountered this SO comment that advises to store it as a `TEXT` (https://stackoverflow.com/a/68541600).

But IMO it is better to be safe than sorry.

If you think we are safe with `BIGINT`, I'll change it no worries.

File: ethtxmanager/sqlstorage/migrations/0001.sql [hunk 14]
```diff
@@ -0,0 +1,31 @@
...
+    gas BIGINT NOT NULL,
+    gas_offset BIGINT,
+    gas_price TEXT,                -- *big.Int
+    blob_sidecar BLOB,
+    blob_gas BIGINT,
+    blob_gas_price TEXT,           -- *big.Int
+    gas_tip_cap TEXT,              -- *big.Int
+    "status" TEXT NOT NULL,
+    block_number TEXT NOT NULL,    -- *big.Int
+    history TEXT,                  -- JSON encoded string
```
[the remark this answers is not in the document]
Comment (Stefan-Ethernal) [reply ethtxmanager/sqlstorage/migrations/0001.sql]: Fixed https://github.com/0xPolygon/zkevm-ethtx-manager/pull/72/commits/f3baefda4f9f064ef01ac2c3a61cdcad9323bcca

File: ethtxmanager/sqlstorage/migrations/0001.sql (hunk 12 above)
[the remark this answers is not in the document]
Comment (Stefan-Ethernal) [reply ethtxmanager/sqlstorage/migrations/0001.sql]: Fixed https://github.com/0xPolygon/zkevm-ethtx-manager/pull/72/commits/f3baefda4f9f064ef01ac2c3a61cdcad9323bcca

File: ethtxmanager/sqlstorage/sqlstorage.go [hunk 15]
```diff
@@ -0,0 +1,271 @@
...
+const (
+	// Insert denotes insert statement
+	Insert sqlAction = iota
+	// Update denotes update statement
+	Update
+	// Delete denotes delete statement
+	Delete
+)
+
+type sqlAction int
```
[the remark this answers is not in the document]
Comment (Stefan-Ethernal) [reply ethtxmanager/sqlstorage/sqlstorage.go]: It is removed, there is no need for it anymore https://github.com/0xPolygon/zkevm-ethtx-manager/pull/72/commits/7a27c787c39c48c908cae92d3226e2f9ab6e8034

File: ethtxmanager/sqlstorage/sqlstorage.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (Stefan-Ethernal) [reply ethtxmanager/sqlstorage/sqlstorage.go]: Fixed, I have returned the `types.AlreadyExists` error, that was already out there https://github.com/0xPolygon/zkevm-ethtx-manager/pull/72/commits/7a27c787c39c48c908cae92d3226e2f9ab6e8034

**Review by joanestebanr — APPROVED** [review]

**Comment by Stefan-Ethernal** [comment]
> I don't known that is done internally but even with just 1 table make sense:

> You do multiples INSERTS but want to rollback if any of them fails
> You do some INSERTS and check something over them in another function (that need to be in the same set of data)
> Basically, it depends how works internally. But in the future make sense, that if we modify it, we think in that model because is the common in the rest of Database usage

I discussed this with @goran-ethernal and @joanestebanr, there is no use case, where we would benefit from introducing transactional model. The transactions are affected one by one.

File: ethtxmanager/sqlstorage/migrations/0001.sql (hunk 11 above)
[the remark this answers is not in the document]
Comment (Stefan-Ethernal) [reply ethtxmanager/sqlstorage/migrations/0001.sql:7]: We have discussed this internally and agreed to remain the `TEXT`. I have only changed `block_number` to be `BIGINT`.

File: ethtxmanager/memory_storage/memstorage.go [hunk 16]
```diff
@@ -0,0 +1,137 @@
+package memorystorage
```
[the remark this answers is not in the document]
Comment (Stefan-Ethernal) [reply ethtxmanager/memory_storage/memstorage.go]: Removed it https://github.com/0xPolygon/zkevm-ethtx-manager/pull/72/commits/eb726498dae62205154404a65648aef6f1af9140

**Review by ToniRamirezM — APPROVED** [review]

File: ethtxmanager/sqlstorage/sqlstorage.go [hunk 17]
```diff
@@ -0,0 +1,287 @@
...
+	"github.com/0xPolygon/zkevm-ethtx-manager/types"
+	"github.com/ethereum/go-ethereum/common"
+	sqlite "github.com/mattn/go-sqlite3"
+	migrate "github.com/rubenv/sql-migrate"
+	"github.com/russross/meddler"
+)
+
+const (
+	// baseSelectQuery represents the base select query, that retrieves all the values from the monitored_txs table
+	baseSelectQuery = `SELECT id, from_address, to_address, nonce, value, tx_data, gas, gas_offset, gas_price,
```
[the remark this answers is not in the document]
Comment (Stefan-Ethernal) [reply ethtxmanager/sqlstorage/sqlstorage.go]: Fixed https://github.com/0xPolygon/zkevm-ethtx-manager/pull/72/commits/c009e714316d1212bdeb4dbd97691a3af38d9716

**Pull request merged**